### ISSUE TITLE
Does not support Apple Silicon

### ISSUE THREAD

**Issue by yaroslavyaroslav** [google/honggfuzz#477]
It fails to build because of there's no build case for Apple silicon.

```log
make
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -c -O3 -mtune=native -funroll-loops -std=c11 -I/usr/local/include -D_GNU_SOURCE -Wall -Wextra -Werror -Wno-format-truncation -Wno-override-init -I. -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.sdk -x objective-c -pedantic -fblocks -Wno-unused-parameter -Wimplicit -Wunused -Wcomment -Wchar-subscripts -Wuninitialized -Wreturn-type -Wpointer-arith -Wno-gnu-case-range -Wno-gnu-designator -Wno-deprecated-declarations -Wno-unknown-pragmas -Wno-attributes -Wno-embedded-directive -Wno-initializer-overrides -Wno-unknown-warning-option -Wno-gnu-empty-initializer -Wno-format-pedantic -Wno-gnu-statement-expression -mllvm -inline-threshold=2000 -D_HF_ARCH_DARWIN -fblocks -o mac/arch.o mac/arch.c
mac/arch.c:699:5: error: unknown type name 'x86_thread_state_t'; did you mean 'arm_thread_state_t'?
    x86_thread_state_t* platform_in_state = ((x86_thread_state_t*)(void*)in_state);
    ^~~~~~~~~~~~~~~~~~
    arm_thread_state_t
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.0.sdk/usr/include/mach/arm/thread_status.h:101:36: note: 'arm_thread_state_t' declared here
typedef _STRUCT_ARM_THREAD_STATE   arm_thread_state_t;
                                   ^
mac/arch.c:699:66: error: expected expression
    x86_thread_state_t* platform_in_state = ((x86_thread_state_t*)(void*)in_state);
                                                                 ^
mac/arch.c:699:47: error: use of undeclared identifier 'x86_thread_state_t'
    x86_thread_state_t* platform_in_state = ((x86_thread_state_t*)(void*)in_state);
                                              ^
mac/arch.c:699:68: error: expected expression
    x86_thread_state_t* platform_in_state = ((x86_thread_state_t*)(void*)in_state);
                                                                   ^
mac/arch.c:701:50: error: no member named 'tsh' in 'struct __darwin_arm_thread_state'
    if (x86_THREAD_STATE32 == platform_in_state->tsh.flavor) {
                              ~~~~~~~~~~~~~~~~~  ^
mac/arch.c:701:9: error: use of undeclared identifier 'x86_THREAD_STATE32'
    if (x86_THREAD_STATE32 == platform_in_state->tsh.flavor) {
        ^
mac/arch.c:702:38: error: no member named 'uts' in 'struct __darwin_arm_thread_state'
        run->pc = platform_in_state->uts.ts32.__eip;
                  ~~~~~~~~~~~~~~~~~  ^
mac/arch.c:704:38: error: no member named 'uts' in 'struct __darwin_arm_thread_state'
        run->pc = platform_in_state->uts.ts64.__rip;
                  ~~~~~~~~~~~~~~~~~  ^
8 errors generated.
make: *** [mac/arch.o] Error 1
```

**Comment by robertswiecki** [comment]
I think this code is used for getting instruction pointer. If your C-fu is strong enough, I'd suggest surrounding it with ifdef's for given CPU arch, like here

https://github.com/google/honggfuzz/blob/master/linux/trace.c#L80

and send me a PR with correct code?

**Comment by yaroslavyaroslav** [comment]
Well my C-fu is quite opposite, actually, but I guess this one I could have handled with some time. For now (totally noob about this codebase) this kind of work looks more boilerplate rather than the one that requires any deep knowledge about that.

**Comment by devnexen** [comment]
> I think this code is used for getting instruction pointer. If your C-fu is strong enough, I'd suggest surrounding it with ifdef's for given CPU arch, like here
> 
> https://github.com/google/honggfuzz/blob/master/linux/trace.c#L80


There is that but ultimately there is also the fact that CrashReport* binaries are for x86 arch ?

**Comment by robertswiecki** [comment]
Yeah, I don't know where to get aarch64 crashwrangler libs - I merely compile it on some mac to verify that it builds, I'm not sure how it exactly works.

FWIW I added code here https://github.com/google/honggfuzz/commit/d9e70def49f628f066df01e473a5ab924d65184c - but one will have to find relevant lib and put it here https://github.com/google/honggfuzz/tree/master/third_party/mac  for it to compile

**Comment by phwd** [comment]
I was under the impression that saelo was able to get it running on an M1 based on https://github.com/googleprojectzero/p0tools/blob/master/iOSOnMac/honggfuzz_ios_on_mac.patch but likewise I cannot find the relevant libs so I'm unsure how it compiled. He stated in his blog

> Furthermore, as honggfuzz itself is still compiled for macOS, some steps of the compilation of the target binary will fail (they will attempt to link previously compiled .o files, but now the platform no longer matches) and so have to be replaced.

https://googleprojectzero.blogspot.com/2021/05/fuzzing-ios-code-on-macos-at-native.html

**Comment by awakened1712** [comment]
> Yeah, I don't know where to get aarch64 crashwrangler libs - I merely compile it on some mac to verify that it builds, I'm not sure how it exactly works.
> 
> FWIW I added code here [d9e70de](https://github.com/google/honggfuzz/commit/d9e70def49f628f066df01e473a5ab924d65184c) - but one will have to find relevant lib and put it here https://github.com/google/honggfuzz/tree/master/third_party/mac for it to compile

Any updates? Do we have a new crashwrangler?

**Comment by devnexen** [comment]
in the "old days" there was the crashwrangler source code if you had a apple license seems no longer the case (I might be wrong) to be able to provide the object file but I would not be surprised if it did not compile on arm64 (again can be wrong here), maybe in the long run the solution is to find a replacement.

**Comment by robertswiecki** [comment]
Could POSIX code work instead?

```OS=POSIX make clean all```

??

**Comment by devnexen** [comment]
wish it did but there is no such timer_t type in macOs.

**Comment by devnexen** [comment]
another failure, libhfcommon/util.c `<link.h>` header does not exist. I ll see at some point if it s doable.

**Comment by raefko** [comment]
Hello, using `OS=POSIX make clean all` I am able to compile.
But when I run the fuzzer with a simple code I have some errors
`[2024-08-02T11:13:15+0200][E][41243] subproc_PrepareExecv():317 lseek(_HF_INPUT_FD=1021, 0, SEEK_SET): Illegal seek
[2024-08-02T11:13:15+0200][E][41243] subproc_New():403 subproc_PrepareExecv() failed
[2024-08-02T11:13:15+0200][W][35940] input_setSize():57 ftruncate(run->dynfile->fd=8, sz=0): Invalid argument
^C[2024-08-02T11:13:15+0200][W][35940] input_setSize():57 ftruncate(run->dynfile->fd=7, sz=0): Invalid argument
[2024-08-02T11:13:15+0200][W][35940] input_setSize():57 ftruncate(run->dynfile->fd=7, sz=8): Invalid argument`

Any fix for this? Thanks !

**Comment by dukecat0** [comment]
Hi, I opened PR #542 to make it easier to use and run on macOS.
